### PR TITLE
osd: set server-side limits on omap get operations

### PR DESCRIPTION
**Pull request by liewegas**
This is important to prevent a silly client from requesting unbounded
work from the OSD (like getting a max of (u64)-1 keys).

### CODE REVIEW

File: PendingReleaseNotes [hunk 1]
```diff
@@ -90,3 +90,12 @@
   jerasure_neon, shec_generic, shec_sse3, shec_sse4, and shec_neon. If you use any
   of these plugins directly you will see a warning in the mon log file.
   Please switch to using just 'jerasure' or 'shec'.
+
+* The librados omap get_keys and get_vals operations include a start key and a
+  limit on the number of keys to return.  The OSD now imposes a configurable
+  limit on the number of keys and number of total bytes it will respond with,
+  which means that a librados user might get fewer keys than they asked for.
+  This is necessary to prevent careless users from requesting and unreasonable
```
Comment (xiexingguo) [inline PendingReleaseNotes]: s/and/an/

File: src/osd/ReplicatedPG.cc [hunk 2]
```diff
@@ -5822,22 +5822,29 @@ int ReplicatedPG::do_osd_ops(OpContext *ctx, vector<OSDOp>& ops)
...
+	  for (num = 0;
+	       num < max_return &&
+		 bl.length() < g_conf->osd_max_omap_bytes_per_request &&
+		 iter->valid();
+	       ++num, iter->next(false)) {
+	    ::encode(iter->key(), bl);
 	  }
 	} // else return empty out_set
-	::encode(out_set, osd_op.outdata);
+	::encode(num, osd_op.outdata);
```
Comment (xiexingguo) [inline src/osd/ReplicatedPG.cc]: I am a little confused here. We change the formatting the result output here. May the consumers still understand it correctly without making any changes

Comment (liewegas) [inline src/osd/ReplicatedPG.cc]: set<string> encodes as a u32 followed by N strings.  We encode the strings in bl above, encode teh count here (once we know it), then append bl.

**Comment by dachary** [comment]
jenkins test this please (eio now ignored on master)

File: src/osd/ReplicatedPG.cc (hunk 2 above)
Comment (xiexingguo) [inline src/osd/ReplicatedPG.cc]: I see, thanks!

**Comment by liewegas** [comment]
@athanatos or @jdurgin mind taking a quick look?

**Comment by athanatos** [comment]
lgtm

**Comment by yuriw** [comment]
http://pulpito.ceph.com/yuriw-2016-11-22_16:38:08-rados-wip-yuri-testing2_2016_11_21-distro-basic-smithi/

http://pulpito.ceph.com/yuriw-2016-11-22_16:38:08-rados-wip-yuri-testing2_2016_11_21-distro-basic-smithi/

**Pull request merged**